libc/sem.cc

```diff
@@ -76,3 +105,81 @@ int sem_trywait(sem_t* s)
         return libc_error(EAGAIN);
     return 0;
 }
+
+//TODO: when are these initialized?
+static std::unordered_map<std::string, indirect_semaphore> named_semaphores;
+static mutex named_semaphores_mutex;
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag){
+    SCOPE_LOCK(named_semaphores_mutex);
+    std::unordered_map<std::string, indirect_semaphore>::iterator iter = 
+                                named_semaphores.find(std::string(name));
+    if(iter != named_semaphores.end()){
+        iter->second.add_reference();
+        return reinterpret_cast<sem_t*>(&(iter->second));
+    }
+    
+    errno = ENOENT;
+    return SEM_FAILED;
+}
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag, mode_t mode, unsigned int value){
+    SCOPE_LOCK(named_semaphores_mutex);
+    std::unordered_map<std::string, indirect_semaphore>::iterator iter = 
+                                named_semaphores.find(std::string(name));
+    
+    if(oflag & O_CREAT){
+        if(iter != named_semaphores.end()){
+            if(oflag & O_EXCL){
+                errno = EEXIST;
+                return SEM_FAILED;
+            }
+
+            iter->second.add_reference();
+            return reinterpret_cast<sem_t*>(&(iter->second));
+        }
+        
+        if(value > SEM_VALUE_MAX){
+            errno = EINVAL;
+            return SEM_FAILED;
+        }
+        /*
+         * Ignoring mode bits bits for now. Do we actually want to worry about 
+         * permissions in a unikernel? Would also need to check permissons when 
+         * using open and unlink if so.
+        */
+        named_semaphores.emplace(std::string(name), indirect_semaphore(value, 1, true)); 
+        return reinterpret_cast<sem_t *>(&named_semaphores[std::string(name)]);
+    }
+
+    errno = ENOENT;
+    return SEM_FAILED;
+}
+
+OSV_LIBC_API
+int sem_unlink(const char *name){
+    SCOPE_LOCK(named_semaphores_mutex);
+    std::unordered_map<std::string, indirect_semaphore>::iterator iter = 
+                                named_semaphores.find(std::string(name));
+    if(iter != named_semaphores.end()){
+        iter->second.unlink();
+        named_semaphores.erase(std::string(name));
+        return 0;
+    }
+    
+    errno = ENOENT;
+    return -1;
+}
+
+OSV_LIBC_API
+int sem_close(sem_t *sem){
+    indirect_semaphore *ind_sem = &from_libc(sem);
+    ind_sem->remove_reference();
+
+    if(!ind_sem->linked() && ind_sem->not_referenced()){
```

Comment:
Should it be "if(ind_sem-linked() ..." instead.

Reply:
From what I can tell, it should be checking ```!ind_sem->linked()``` as opposed to ```ind_sem->linked()```. This way, the underlying semaphore will only be destroyed when it is no longer possible to create links to it (after having been unlinked).

I have been working based off of these descriptions of [sem_unlink](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_unlink.html) and [sem_close](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_close.html) from the IEEE standard. Particularly 

>"If the semaphore has not been removed with a successful call to [sem_unlink()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_unlink.html), then sem_close() has no effect on the state of the semaphore. If the [sem_unlink()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_unlink.html) function has been successfully invoked for name after the most recent call to [sem_open()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_open.html) with O_CREAT for this semaphore, then when all processes that have opened the semaphore close it, the semaphore is no longer accessible."

would seem to me to suggest the current implementation.

Reply:
@nyh Can we resolve it as my original comment was in the context of the outdate code now? And @StonewallJohnson fixed it the way you wanted it?